index.js

```diff
@@ -164,124 +157,136 @@ function onRequestResponse (options, error, response, body) {
+    return onChallenge(options, response, stringBody);
   }
 
   // All is good
-  processResponseBody(options, response, body);
+  onRequestComplete(options, response, body);
 }
 
 function validate (options, response, body) {
-  var match;
+  let match;
 
   // Finding captcha
   if (body.indexOf('why_captcha') !== -1 || /cdn-cgi\/l\/chk_captcha/i.test(body)) {
-    throw new errors.CaptchaError('captcha', options, response);
+    // Convenience boolean
+    response.isCaptcha = true;
+    throw new CaptchaError('captcha', options, response);
   }
 
   // Trying to find '<span class="cf-error-code">1006</span>'
   match = body.match(/<\w+\s+class="cf-error-code">(.*)<\/\w+>/i);
 
   if (match) {
-    var code = parseInt(match[1]);
-    throw new errors.CloudflareError(code, options, response);
+    let code = parseInt(match[1]);
+    throw new CloudflareError(code, options, response);
   }
 
   return false;
 }
 
-function solveChallenge (options, response, body) {
-  var callback = options.callback;
-  var cause;
-  var error;
+function onChallenge (options, response, body) {
+  const callback = options.callback;
+  const uri = response.request.uri;
+  // The query string to send back to Cloudflare
+  const payload = { /* s, jschl_vc, pass, jschl_answer */ };
+
+  let cause;
+  let error;
 
   if (options.challengesToSolve === 0) {
     cause = 'Cloudflare challenge loop';
-    error = new errors.CloudflareError(cause, options, response);
+    error = new CloudflareError(cause, options, response);
     error.errorType = 4;
 
     return callback(error);
   }
 
-  var timeout = parseInt(options.cloudflareTimeout);
-  var uri = response.request.uri;
-  // The query string to send back to Cloudflare
-  // var payload = { s, jschl_vc, pass, jschl_answer };
-  var payload = {};
-  var sandbox;
-  var match;
+  let timeout = parseInt(options.cloudflareTimeout);
+  let match;
 
   match = body.match(/name="s" value="(.+?)"/);
-
   if (match) {
     payload.s = match[1];
   }
 
   match = body.match(/name="jschl_vc" value="(\w+)"/);
-
   if (!match) {
     cause = 'challengeId (jschl_vc) extraction failed';
-    return callback(new errors.ParserError(cause, options, response));
+    return callback(new ParserError(cause, options, response));
   }
 
   payload.jschl_vc = match[1];
 
   match = body.match(/name="pass" value="(.+?)"/);
-
   if (!match) {
     cause = 'Attribute (pass) value extraction failed';
-    return callback(new errors.ParserError(cause, options, response));
+    return callback(new ParserError(cause, options, response));
   }
 
   payload.pass = match[1];
 
   match = body.match(/getElementById\('cf-content'\)[\s\S]+?setTimeout.+?\r?\n([\s\S]+?a\.value\s*=.+?)\r?\n(?:[^{<>]*},\s*(\d{4,}))?/);
-
   if (!match) {
     cause = 'setTimeout callback extraction failed';
-    return callback(new errors.ParserError(cause, options, response));
+    return callback(new ParserError(cause, options, response));
   }
 
   if (isNaN(timeout)) {
```

Comment:
The match length will always be 2 if there's a match. This should be changed to `if (match[2] !== undefined)`. I didn't modify this in this PR. Just noting it here.

Reply:
This is not resolved, just marking as such since the PR doesn't modify this part of the coe.